types/react-query/index.d.ts

```diff
@@ -36,107 +53,201 @@ export interface QueryOptions<TResult> {
+    refetch: (options?: {
+        force?: boolean;
+        thrownOnError?: boolean;
+    }) => void;
 }
 
-export interface QueryResultPaginated<TResult, TVariables> extends QueryResult<TResult[], TVariables> {
-    isFetchingMore: boolean;
-    canFetchMore: boolean;
-    fetchMore: (variables?: TVariables) => Promise<TResult>;
+export type PaginatedQueryState<TResult> =
+    | PaginatedQueryLoadingState
+    | PaginatedQueryErrorState
+    | PaginatedQuerySuccessState<TResult>;
+
+export interface PaginatedQueryLoadingState extends QueryBaseState {
+    status: 'loading';
+    resolvedData: undefined;
+    latestData: undefined;
+    error: null;
 }
 
-export function prefetchQuery<TResult, TVariables extends object>(
-    queryKey: QueryKey<TVariables>,
-    queryFn: QueryFunction<TResult, TVariables>,
-    options?: PrefetchQueryOptions<TResult>
-): Promise<TResult>;
+export interface PaginatedQueryErrorState extends QueryBaseState {
+    status: 'error';
+    resolvedData: undefined;
+    latestData: undefined;
+    error: Error;
+}
+
+export interface PaginatedQuerySuccessState<TResult> extends QueryBaseState {
+    status: 'success';
+    resolvedData: TResult;
+    latestData: TResult;
+    error: null;
+}
+
+export type InfiniteQueryState<TResult, TVariable> =
+    | InfiniteQuerySuccessState<TResult, TVariable>
+    | InfiniteQueryErrorState<TVariable>
+    | InfiniteQueryLoadingState<TVariable>;
+
+export interface InfiniteQueryLoadingState<TVariable> extends InfiniteQueryBaseState<TVariable> {
+    status: 'loading';
+    data: [];
+    error: null;
+}
+
+export interface InfiniteQueryErrorState<TVariable> extends InfiniteQueryBaseState<TVariable> {
+    status: 'error';
+    data: [];
+    error: Error;
+}
 
-export interface PrefetchQueryOptions<TResult> extends QueryOptions<TResult> {
-    force?: boolean;
+export interface InfiniteQuerySuccessState<TResult, TVariable> extends InfiniteQueryBaseState<TVariable> {
+    status: 'success';
+    data: TResult[];
+    error: null;
+}
+
+export interface InfiniteQueryBaseState<TVariable> {
+    isFetching: boolean;
+    isFetchingMore: boolean;
+    failureCount: number;
+    refetch: (options?: {
+        force?: boolean;
+        thrownOnError?: boolean;
+    }) => void;
+    fetchMore: (variables?: TVariable) => Promise<any>;
+    canFetchMore: boolean;
 }
 
 export function useMutation<TResults, TVariables extends object>(
     mutationFn: MutationFunction<TResults, TVariables>,
-    mutationOptions?: MutationOptions
-): [MutateFunction<TResults, TVariables>, MutationResult<TResults>];
+    mutationOptions?: MutationOptions<TResults>,
+): [MutateFunction<TResults, TVariables>, MutationState<TResults>];
 
-export type MutationFunction<TResults, TVariables extends object> = (
-    variables: TVariables,
-) => Promise<TResults>;
+export type MutationFunction<TResults, TVariables extends object> = (variables: TVariables) => Promise<TResults>;
```

Comment:
Why `TVariables extends object`? Can't `variables` also be a primitive type like `string` or `number`?

Reply:
it should be a generic with a reasonable default

Reply:
And also the variables are always passed as object to the mutation function so it has to extend object.

Reply:
Definition of `useMutation` was mostly copy-pasted from v.0, but I guess it need some attention too!

Reply:
Fixed. However, I had trouble correctly typing first argument of mutate function - see TODOs in test file. Any help would be appreciated